snuba/request/request_settings.py

```diff
@@ -36,6 +92,10 @@ def get_rate_limit_params(self) -> Sequence[RateLimitParameters]:
     def add_rate_limit(self, rate_limit_param: RateLimitParameters) -> None:
         pass
 
+    @abstractmethod
+    def get_sampling_config(self) -> SamplingConfig:
+        pass
+
```

Comment:
I went back and forth quite a bit on whether to add the sampling config in the request settings or to the Query object.
It seems appropriate to add it to the settings in that it is something immutable provided by the user through the request, that is not part of the query but drives how the query is built.
Considering the actual sampling rate is defined through query processing (whether in snuba query processing or clickhouse query processing. The point still holds), putting this field in the Query object (instead of in settings) would mean that, after query processing, that object would include both the decided sampling rate and the directive provided by the client to decide it, which, instead, should not be available anymore.

The only downside to storing sampling config in request settings, is that it may, in the future, become a problem for split queries. Request settings should be immutable in that part of the request, but, potentially we may want to apply a different sampling strategy to parts of a split query.
Anyway the current split query does not have this problem (all the components the query is split into should have the same sampling config) and we do not have a clear idea of what kind of splits we are going to support beyond the two cases we know about. So it seems premature for split query to drive the decision around sampling_config.

Reply:
> Considering the actual sampling rate is defined through query processing (whether in snuba query processing or clickhouse query processing. The point still holds), putting this field in the Query object (instead of in settings) would mean that, after query processing, that object would include both the decided sampling rate and the directive provided by the client to decide it, which, instead, should not be available anymore.

No strong opinion on this at the moment. Settings seems fine (my opinion might change later as I better understand what's happening here.)

> The only downside to storing sampling config in request settings, is that it may, in the future, become a problem for split queries. Request settings should be immutable in that part of the request, but, potentially we may want to apply a different sampling strategy to parts of a split query.

I'm struggling to identify a case where this would be beneficial. I do not think we should design for this.

> Anyway the current split query does not have this problem (all the components the query is split into should have the same sampling config)

This seems more incidental than intentional.

> and we do not have a clear idea of what kind of splits we are going to support beyond the two cases we know about. So it seems premature for split query to drive the decision around sampling_config.

I strongly disagree with this perspective. Both of splitting and sampling are complex on their own and the combined behavior is even more complex, and I think not intentionally setting boundaries and expectations about how they relate does more harm than good by introducing a lot of ambiguity. (See review comment.)

Reply:
See my long comment about the relationship between split and sampling.